File: Test-Corpus/claude-cross-prompt/claude-opus-5-1m/B02_synthetic/hashmap-tree/translated/c_src/src/hashmap.c
/*
 * Copyright 2025 MIT Lincoln Laboratory
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software
 * and associated documentation files (the "Software"),
 * to deal in the Software without restriction,
 * including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice
 * shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
 * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
// hashmap.c
#include "hashmap.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

static uint64_t hash_function(tree_id_t key) {
    // FNV-1a hash
    uint64_t hash = 14695981039346656037ULL;
    uint8_t *bytes = (uint8_t *)&key;
    
    for (size_t i = 0; i < sizeof(tree_id_t); i++) {
        hash ^= bytes[i];
        hash *= 1099511628211ULL;
    }
    
    return hash;
}

static int should_resize(hashmap_t *map) {
    double load = (double)(map->size + map->deleted_count) / map->capacity;
    return load > HASHMAP_LOAD_FACTOR;
}

static int hashmap_resize(hashmap_t *map) {
    size_t old_capacity = map->capacity;
    hashmap_entry_t *old_entries = map->entries;
    
    // Double capacity
    map->capacity *= 2;
    map->entries = calloc(map->capacity, sizeof(hashmap_entry_t));
    if (!map->entries) {
        map->entries = old_entries;
        map->capacity = old_capacity;
        return -1;
    }
    
    map->size = 0;
    map->deleted_count = 0;
    
    // Rehash all entries
    for (size_t i = 0; i < old_capacity; i++) {
        if (old_entries[i].occupied && !old_entries[i].deleted) {
            hashmap_put(map, old_entries[i].key, old_entries[i].value);
        }
    }
    
    free(old_entries);
    return 0;
}

hashmap_t* hashmap_create(void) {
    hashmap_t *map = malloc(sizeof(hashmap_t));
    if (!map) {
        return NULL;
    }
    
    map->capacity = HASHMAP_INITIAL_CAPACITY;
    map->size = 0;
    map->deleted_count = 0;
    map->entries = calloc(map->capacity, sizeof(hashmap_entry_t));
    
    if (!map->entries) {
        free(map);
        return NULL;
    }
    
    return map;
}

void hashmap_destroy(hashmap_t *map) {
    if (map) {
        free(map->entries);
        free(map);
    }
}

int hashmap_put(hashmap_t *map, tree_id_t key, void *value) {
    if (!map) {
        return -1;
    }
    
    if (should_resize(map)) {
        if (hashmap_resize(map) != 0) {
            return -1;
        }
    }
    
    uint64_t hash = hash_function(key);
    size_t index = hash % map->capacity;
    size_t probe = 0;
    
    // Linear probing
    while (probe < map->capacity) {
        size_t current = (index + probe) % map->capacity;
        
        if (!map->entries[current].occupied) {
            // Empty slot
            map->entries[current].key = key;
            map->entries[current].value = value;
            map->entries[current].occupied = 1;
            map->entries[current].deleted = 0;
            map->size++;
            return 0;
        } else if (map->entries[current].deleted) {
            // Reuse deleted slot
            map->entries[current].key = key;
            map->entries[current].value = value;
            map->entries[current].deleted = 0;
            map->size++;
            map->deleted_count--;
            return 0;
        } else if (map->entries[current].key == key) {
            // Update existing
            map->entries[current].value = value;
            return 0;
        }
        
        probe++;
    }
    
    return -1; // Map is full (shouldn't happen with resizing)
}

void* hashmap_get(hashmap_t *map, tree_id_t key) {
    if (!map) {
        return NULL;
    }
    
    uint64_t hash = hash_function(key);
    size_t index = hash % map->capacity;
    size_t probe = 0;
    
    while (probe < map->capacity) {
        size_t current = (index + probe) % map->capacity;
        
        if (!map->entries[current].occupied) {
            return NULL;
        }
        
        if (!map->entries[current].deleted && map->entries[current].key == key) {
            return map->entries[current].value;
        }
        
        probe++;
    }
    
    return NULL;
}

void* hashmap_remove(hashmap_t *map, tree_id_t key) {
    if (!map) {
        return NULL;
    }
    
    uint64_t hash = hash_function(key);
    size_t index = hash % map->capacity;
    size_t probe = 0;
    
    while (probe < map->capacity) {
        size_t current = (index + probe) % map->capacity;
        
        if (!map->entries[current].occupied) {
            return NULL;
        }
        
        if (!map->entries[current].deleted && map->entries[current].key == key) {
            void *value = map->entries[current].value;
            map->entries[current].deleted = 1;
            map->size--;
            map->deleted_count++;
            return value;
        }
        
        probe++;
    }
    
    return NULL;
}

int hashmap_contains(hashmap_t *map, tree_id_t key) {
    return hashmap_get(map, key) != NULL;
}

size_t hashmap_size(hashmap_t *map) {
    return map ? map->size : 0;
}

void hashmap_clear(hashmap_t *map) {
    if (!map) {
        return;
    }
    
    for (size_t i = 0; i < map->capacity; i++) {
        map->entries[i].occupied = 0;
        map->entries[i].deleted = 0;
    }
    
    map->size = 0;
    map->deleted_count = 0;
}

File: Test-Corpus/claude-minimal/claude-opus-5-1m/B02_synthetic/hashmap-tree/translated/src/hashmap.rs
// hashmap.rs - Rust translation of hashmap.c/.h
use std::ffi::c_void;
use std::ptr;

pub type TreeId = u64;

pub const HASHMAP_INITIAL_CAPACITY: usize = 16;
pub const HASHMAP_LOAD_FACTOR: f64 = 0.75;

#[derive(Clone, Copy)]
pub struct HashmapEntry {
    pub key: TreeId,
    pub value: *mut c_void,
    pub occupied: bool,
    pub deleted: bool,
}

impl Default for HashmapEntry {
    fn default() -> Self {
        HashmapEntry {
            key: 0,
            value: ptr::null_mut(),
            occupied: false,
            deleted: false,
        }
    }
}

pub struct Hashmap {
    pub entries: Vec<HashmapEntry>,
    pub capacity: usize,
    pub size: usize,
    pub deleted_count: usize,
}

fn hash_function(key: TreeId) -> u64 {
    // FNV-1a hash
    let mut hash: u64 = 14695981039346656037;
    let bytes = key.to_ne_bytes();
    for b in bytes.iter() {
        hash ^= *b as u64;
        hash = hash.wrapping_mul(1099511628211);
    }
    hash
}

pub fn hashmap_create() -> Box<Hashmap> {
    let cap = HASHMAP_INITIAL_CAPACITY;
    Box::new(Hashmap {
        entries: vec![HashmapEntry::default(); cap],
        capacity: cap,
        size: 0,
        deleted_count: 0,
    })
}

pub fn hashmap_destroy(map: Box<Hashmap>) {
    drop(map);
}

fn should_resize(map: &Hashmap) -> bool {
    let load = (map.size + map.deleted_count) as f64 / map.capacity as f64;
    load > HASHMAP_LOAD_FACTOR
}

fn hashmap_resize(map: &mut Hashmap) -> i32 {
    let old_capacity = map.capacity;
    let old_entries = std::mem::take(&mut map.entries);

    map.capacity = old_capacity * 2;
    map.entries = vec![HashmapEntry::default(); map.capacity];
    map.size = 0;
    map.deleted_count = 0;

    for i in 0..old_capacity {
        if old_entries[i].occupied && !old_entries[i].deleted {
            hashmap_put(map, old_entries[i].key, old_entries[i].value);
        }
    }
    0
}

pub fn hashmap_put(map: &mut Hashmap, key: TreeId, value: *mut c_void) -> i32 {
    if should_resize(map) {
        if hashmap_resize(map) != 0 {
            return -1;
        }
    }

    let hash = hash_function(key);
    let index = (hash as usize) % map.capacity;

    for probe in 0..map.capacity {
        let current = (index + probe) % map.capacity;

        if !map.entries[current].occupied {
            map.entries[current].key = key;
            map.entries[current].value = value;
            map.entries[current].occupied = true;
            map.entries[current].deleted = false;
            map.size += 1;
            return 0;
        } else if map.entries[current].deleted {
            map.entries[current].key = key;
            map.entries[current].value = value;
            map.entries[current].deleted = false;
            map.size += 1;
            map.deleted_count -= 1;
            return 0;
        } else if map.entries[current].key == key {
            map.entries[current].value = value;
            return 0;
        }
    }
    -1
}

pub fn hashmap_get(map: &Hashmap, key: TreeId) -> *mut c_void {
    let hash = hash_function(key);
    let index = (hash as usize) % map.capacity;

    for probe in 0..map.capacity {
        let current = (index + probe) % map.capacity;
        if !map.entries[current].occupied {
            return ptr::null_mut();
        }
        if !map.entries[current].deleted && map.entries[current].key == key {
            return map.entries[current].value;
        }
    }
    ptr::null_mut()
}

pub fn hashmap_remove(map: &mut Hashmap, key: TreeId) -> *mut c_void {
    let hash = hash_function(key);
    let index = (hash as usize) % map.capacity;

    for probe in 0..map.capacity {
        let current = (index + probe) % map.capacity;
        if !map.entries[current].occupied {
            return ptr::null_mut();
        }
        if !map.entries[current].deleted && map.entries[current].key == key {
            let value = map.entries[current].value;
            map.entries[current].deleted = true;
            map.size -= 1;
            map.deleted_count += 1;
            return value;
        }
    }
    ptr::null_mut()
}

pub fn hashmap_contains(map: &Hashmap, key: TreeId) -> bool {
    !hashmap_get(map, key).is_null()
}

pub fn hashmap_size(map: &Hashmap) -> usize {
    map.size
}

pub fn hashmap_clear(map: &mut Hashmap) {
    for entry in map.entries.iter_mut() {
        entry.occupied = false;
        entry.deleted = false;
    }
    map.size = 0;
    map.deleted_count = 0;
}
